File: python/mcp_server/tool_registry.py
"""Tool registry for MCP server."""

from __future__ import annotations

import json
from typing import Any, Callable, Dict, List, Optional

from pydantic import BaseModel


class ToolSchema(BaseModel):
    """MCP tool schema."""

    name: str
    description: str
    inputSchema: Dict[str, Any]


class ToolRegistry:
    """Central registry for MCP tools."""

    def __init__(self):
        self._tools: Dict[str, Callable] = {}
        self._schemas: Dict[str, ToolSchema] = {}

    def register_tool(
        self,
        name: str,
        description: str,
        input_schema: Dict[str, Any],
        handler: Callable,
    ) -> None:
        """Register a tool."""
        self._tools[name] = handler
        self._schemas[name] = ToolSchema(
            name=name, description=description, inputSchema=input_schema
        )

    def list_tools(self) -> List[Dict[str, Any]]:
        """List all registered tools."""
        return [
            {
                "name": schema.name,
                "description": schema.description,
                "inputSchema": schema.inputSchema,
            }
            for schema in self._schemas.values()
        ]

    def get_tool(self, name: str) -> Optional[Callable]:
        """Get tool handler by name."""
        return self._tools.get(name)

    def call_tool(self, name: str, arguments: Dict[str, Any]) -> Dict[str, Any]:
        """Call a tool by name with arguments."""
        handler = self._tools.get(name)
        if not handler:
            raise ValueError(f"Tool not found: {name}")

        try:
            result = handler(**arguments)
            if isinstance(result, dict):
                return result
            return {"content": [{"type": "text", "text": json.dumps(result)}]}
        except Exception as e:
            return {
                "content": [{"type": "text", "text": f"Error: {str(e)}"}],
                "isError": True,
            }

File: openclaw-plugin/src/tools/tool-registry.ts
/**
 * Tool Registry for managing OpenClaw tools
 */

import { ToolSchema, ToolHandler, ToolResponse } from '../types';

export class ToolRegistry {
  private tools = new Map<string, ToolSchema & { handler: ToolHandler }>();

  /**
   * Register a tool with the registry
   */
  register(tool: ToolSchema & { handler: ToolHandler }): void {
    if (this.tools.has(tool.name)) {
      throw new Error(`Tool ${tool.name} is already registered`);
    }
    this.tools.set(tool.name, tool);
  }

  /**
   * Get all registered tools
   */
  getAll(): Array<ToolSchema & { handler: ToolHandler }> {
    return Array.from(this.tools.values());
  }

  /**
   * Get a specific tool by name
   */
  get(name: string): (ToolSchema & { handler: ToolHandler }) | undefined {
    return this.tools.get(name);
  }

  /**
   * Check if a tool exists
   */
  has(name: string): boolean {
    return this.tools.has(name);
  }

  /**
   * Call a tool by name with arguments
   */
  async call(name: string, args: Record<string, any>): Promise<ToolResponse> {
    const tool = this.get(name);
    if (!tool) {
      throw new Error(`Tool ${name} not found`);
    }
    return tool.handler(args);
  }
}
